lib/Runtime/Library/JavascriptExceptionMetadata.cpp

```diff
@@ -0,0 +1,144 @@
+            scriptContext);
+        Js::JavascriptOperators::OP_SetProperty(metadata, Js::PropertyIds::source, var, scriptContext);
+
+        var = Js::JavascriptConversion::ToString(
+            Js::JavascriptOperators::OP_GetProperty(exception, Js::PropertyIds::url, scriptContext),
+            scriptContext);
+        Js::JavascriptOperators::OP_SetProperty(metadata, Js::PropertyIds::url, var, scriptContext);
+    }
+
+    bool JavascriptExceptionMetadata::PopulateMetadataFromException(Var metadata, JavascriptExceptionObject * recordedException, ScriptContext * scriptContext) {
+        uint32 offset = recordedException->GetByteCodeOffset();
+        FunctionBody * functionBody = recordedException->GetFunctionBody();
+        ULONG line;
+        LONG column;
+        if (functionBody->GetUtf8SourceInfo()->GetIsLibraryCode() ||
+            !functionBody->GetLineCharOffset(offset, &line, &column)) {
+            line = 0;
+            column = 0;
+        }
+
+        Js::JavascriptOperators::OP_SetProperty(metadata, Js::PropertyIds::line,
+            Js::JavascriptNumber::New(line, scriptContext), scriptContext);
+
+        Js::JavascriptOperators::OP_SetProperty(metadata, Js::PropertyIds::column,
+            Js::JavascriptNumber::New(column, scriptContext), scriptContext);
+
+        Js::JavascriptOperators::OP_SetProperty(metadata, Js::PropertyIds::length,
+            Js::JavascriptNumber::New(0, scriptContext), scriptContext);
+
+        Js::JavascriptOperators::OP_SetProperty(metadata, Js::PropertyIds::url,
+            Js::JavascriptString::NewCopySz(functionBody->GetSourceContextInfo()->url, scriptContext), scriptContext);
+
+        Js::Utf8SourceInfo* sourceInfo = functionBody->GetUtf8SourceInfo();
+        sourceInfo->EnsureLineOffsetCache();
+        JsUtil::LineOffsetCache<Recycler> *cache = sourceInfo->GetLineOffsetCache();
+
+        LPCUTF8 functionSource = sourceInfo->GetSource(_u("Jsrt::JsExperimentalGetAndClearExceptionWithMetadata"));
+
+        charcount_t startByteOffset;
+        charcount_t endByteOffset;
+        charcount_t startCharOffset;
+        charcount_t endCharOffset;
+
+        if (line >= cache->GetLineCount())
+        {
+            return false;
+        }
+        startCharOffset = cache->GetCharacterOffsetForLine(line, &startByteOffset);
+
+        if (line + 1 >= cache->GetLineCount())
+        {
+            endByteOffset = functionBody->LengthInBytes();
+            endCharOffset = functionBody->LengthInChars();
+        }
+        else
+        {
+            endCharOffset = cache->GetCharacterOffsetForLine(line + 1, &endByteOffset);
+
+            // The offsets above point to the start of the following line,
+            // while we need to find the end of the current line.
+            // To do so, just step back over the preceeding newline character(s)
+            if (functionSource[endByteOffset - 1] == _u('\n'))
```

Comment:
This must not be 0 since it is the offset for at least the second line. However, the next check could potentially be 0, so I've added a check there.